ASSERTION FAILED: InjectedBundle::singleton().isTestRunning() is constantly crashing fast/eventsource/eventsource-attribute-listeners.html

aprotyas opened the pull request:
#### ad80cae882970a64828991d1848c225906493176
<pre>
ASSERTION FAILED: InjectedBundle::singleton().isTestRunning() is constantly crashing fast/eventsource/eventsource-attribute-listeners.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=297141">https://bugs.webkit.org/show_bug.cgi?id=297141</a>
<a href="https://rdar.apple.com/157879724">rdar://157879724</a>

Reviewed by Alexey Proskuryakov.

fast/eventsource/eventsource-attribute-listeners.html is crashing in
EWS, not because of any fault of its own, but because it usually runs in
succession to fast/events/touch/page-scaled-touch-gesture-click.html.

The bad test calls an asynchronous function in the body onload handler,
i.e. does not await. This can put the test infrastructure in a bad
state. We receive InjectedBundle::done() early because the test is so
quickly done with and as a result the injected bundle clears out its
held TestRunner instance. However, TestRunner has a positive refcount
because of InjectedBundlePage::frameDidChangeLocation() --&gt; dump(). We
end up in this code path because the test runner is _not_ currently
&quot;waiting for done&quot;, which we should have, had the test properly run its
runTest() method. Finally, when the runTest() method does eventually
finish executing, we reach TestRunner::waitUntilDone() (since it still
has a positive refcount), and crash at the release assert because the
injected bundle has already cleared out its test runner instance.

In the good case, we never end up having a high refcount through dump()
because the test runner would have been &quot;waiting for done&quot; already, but
more importantly, we don&apos;t think the test is done because the async test
code is actually awaited on. As a result, we don&apos;t accidentally wipe the
test runner instance held by the injected bundle singleton.

While the TestRunner being kept alive may seem confusing, the ref counts
are never unbalanced, and it&apos;s fair for TestRunner callers to keep it
alive. What&apos;s unfortunate is that we go against the standard practice
from elsewhere in TestRunner and just release assert over the absence of
a test runner instance in the injected bundle. Everywhere else just
early returns from the work it is supposed to do. This patch follows
suit so that tests do not unnecessarily crash if they&apos;re poorly written.
We do add an error log to surface some information.

To address the test&apos;s badness, we first extricate the async function call
into a proper &apos;load&apos; event handler. The test still times out, though, as
we already know in <a href="https://rdar.apple.com/80387340">rdar://80387340</a>. This patch also does the bare minimum
to move the test from a time out to a consistent failure by moving away
from unuspported EventSender API.

Finally, we make two test expectation changes:

1. Remove the crashing expectation for fast/eventsource/eventsource-attribute-listeners.html
2. Remove the &apos;failing&apos; expectation because fast/events/touch/page-scaled-touch-gesture-click.html
   still relies on touch event support which is missing from open source
   configurations, and hence should be skipped altogether, like all the
   other tests under fast/events/touch.

* LayoutTests/fast/events/touch/page-scaled-touch-gesture-click.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::waitUntilDone):

Canonical link: <a href="https://commits.webkit.org/299084@main">https://commits.webkit.org/299084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30bbdd9fe2416d8eae7159983f404565973a2775

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117742 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28050 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123876 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69756 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/865b683f-4c0d-49a4-953d-1eb5a2a5a8a8) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119620 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38112 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46002 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89361 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/56206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9b22df6d-af11-4302-ad53-226472fdf43e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120694 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105573 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69854 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/cache (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/83ff3f62-a834-43d6-af43-e28e616f2de2) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29411 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23688 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67535 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99760 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23868 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126969 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44645 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33602 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98016 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45003 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101798 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97804 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43183 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21153 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41010 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18794 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44517 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43975 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47322 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45666 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/54a7b10beae4810e04eaeb9f548df5e0d9ee81ba. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49709

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1e58bb6af21847e1c3b7b16198bad5adf52d92a0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49709

aprotyas commented:
> Thank you for the fix! Is it possible to change something in WebKitTestRunner to prevent a test ever breaking another test like this?

@aproskuryakov yes, done, I looked at this again and was able to make sense of what the bad state looks like.

aprotyas commented on Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
```diff
@@ -225,7 +225,10 @@ void TestRunner::waitUntilDownloadFinished()
 {
-    RELEASE_ASSERT(InjectedBundle::singleton().isTestRunning());
+    if (!InjectedBundle::singleton().isTestRunning()) {
+        LOG_ERROR("testRunner.waitUntilDone() called after test has terminated. Possibly an async handler was not awaited.");
```
No it wouldn't, I don't think so. This is the waitUntilDone call for the bad test, which we won't respect.

aproskuryakov replied:
Hmm. But the bad test (fast/events/touch/page-scaled-touch-gesture-click.html) wasn't crashing, meaning that `InjectedBundle::singleton().isTestRunning()` was always true in it, meaning that this patch changes nothing about what happens in the bad test.

aproskuryakov replied:
Perhaps you are saying that this is `TestRunner` for the bad test? But the problem happened when fast/eventsource/eventsource-attribute-listeners.html was considered running, so the print will be associated with the victim test.

aprotyas replied:
Oh, yes, I see what you mean. I think the LOG_ERROR might be a bad signal

aprotyas replied:
I was just thinking, since the test runner instance is always associated with the bad test (even though the crash occurs at a later time that makes our tooling attribute it to a different test), I can just include the faulty test URL in the LOG_ERROR statement.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/30bbdd9fe2416d8eae7159983f404565973a2775. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49709

aprotyas committed:
ASSERTION FAILED: InjectedBundle::singleton().isTestRunning() is constantly crashing fast/eventsource/eventsource-attribute-listeners.html
https://bugs.webkit.org/show_bug.cgi?id=297141
rdar://157879724

Reviewed by Alexey Proskuryakov.

fast/eventsource/eventsource-attribute-listeners.html is crashing in
EWS, not because of any fault of its own, but because it usually runs in
succession to fast/events/touch/page-scaled-touch-gesture-click.html.

The bad test calls an asynchronous function in the body onload handler,
i.e. does not await. This can put the test infrastructure in a bad
state. We receive InjectedBundle::done() early because the test is so
quickly done with and as a result the injected bundle clears out its
held TestRunner instance. However, TestRunner has a positive refcount
because of InjectedBundlePage::frameDidChangeLocation() --> dump(). We
end up in this code path because the test runner is _not_ currently
"waiting for done", which we should have, had the test properly run its
runTest() method. Finally, when the runTest() method does eventually
finish executing, we reach TestRunner::waitUntilDone() (since it still
has a positive refcount), and crash at the release assert because the
injected bundle has already cleared out its test runner instance.

In the good case, we never end up having a high refcount through dump()
because the test runner would have been "waiting for done" already, but
more importantly, we don't think the test is done because the async test
code is actually awaited on. As a result, we don't accidentally wipe the
test runner instance held by the injected bundle singleton.

While the TestRunner being kept alive may seem confusing, the ref counts
are never unbalanced, and it's fair for TestRunner callers to keep it
alive. What's unfortunate is that we go against the standard practice
from elsewhere in TestRunner and just release assert over the absence of
a test runner instance in the injected bundle. Everywhere else just
early returns from the work it is supposed to do. This patch follows
suit so that tests do not unnecessarily crash if they're poorly written.
We do add an error log to surface some information.

To address the test's badness, we first extricate the async function call
into a proper 'load' event handler. The test still times out, though, as
we already know in rdar://80387340. This patch also does the bare minimum
to move the test from a time out to a consistent failure by moving away
from unuspported EventSender API.

Finally, we make two test expectation changes:

1. Remove the crashing expectation for fast/eventsource/eventsource-attribute-listeners.html
2. Remove the 'failing' expectation because fast/events/touch/page-scaled-touch-gesture-click.html
   still relies on touch event support which is missing from open source
   configurations, and hence should be skipped altogether, like all the
   other tests under fast/events/touch.

* LayoutTests/fast/events/touch/page-scaled-touch-gesture-click.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::waitUntilDone):

Canonical link: https://commits.webkit.org/299084@main
